File: Android-MVVM-Java/app/src/main/java/com/mvvm/java/core/base/mvvm/MVVMActivity.java
package com.mvvm.java.core.base.mvvm;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.viewbinding.ViewBinding;

import com.mvvm.java.core.base.BaseApplication;
import com.mvvm.java.core.base.BaseMvvmActivity;
import com.mvvm.java.core.base.BaseMvvmView;
import com.mvvm.java.core.base.BaseMvvmViewModel;
import com.mvvm.java.core.di.component.ActivityComponent;
import com.mvvm.java.core.di.component.DaggerActivityComponent;
import com.mvvm.java.core.di.module.ActivityModule;

import javax.inject.Inject;

public abstract class MVVMActivity<VB extends ViewBinding, VM extends BaseMvvmViewModel>
        extends BaseMvvmActivity<VB> implements BaseMvvmView {

    @Inject
    public VM viewModel;

    private ActivityComponent activityComponent;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityComponent = DaggerActivityComponent.builder()
                .activityModule(new ActivityModule(this))
                .applicationComponent(((BaseApplication) getApplication()).getComponent())
                .build();
    }

    public ActivityComponent getActivityComponent() {
        return activityComponent;
    }
}

File: android-mvvm-kotlin/app/src/main/java/com/base/mvvm/core/base/mvvm/MVVMActivity.kt
package com.base.mvvm.core.base.mvvm

import android.os.Bundle
import androidx.databinding.ViewDataBinding
import com.base.mvvm.core.base.BaseMVVMActivity
import com.base.mvvm.core.base.BaseViewModel

/**
 * Author: William Giang Nguyen | 8/7/2022
 * */
abstract class MVVMActivity<BD : ViewDataBinding, VM : BaseViewModel> :
    BaseMVVMActivity<BD>() {

    private lateinit var viewModel: VM

    abstract fun getVM(): VM

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        viewModel = getVM()
    }

}